src/ansi-c/c_typecheck_base.cpp

```diff
@@ -744,6 +760,48 @@ void c_typecheck_baset::typecheck_declaration(
           if(return_type.id() != ID_empty)
             parameter_map.erase(CPROVER_PREFIX "return_value");
         }
+
+        // create a contract symbol
+        symbolt contract;
+        contract.name = "contract::" + id2string(new_symbol.name);
+        contract.base_name = new_symbol.name;
+        contract.pretty_name = new_symbol.name;
+        contract.is_property = true;
+        contract.type = code_type;
+        contract.mode = new_symbol.mode;
+        contract.module = module;
+        contract.location = new_symbol.location;
+
+        auto entry = symbol_table.insert(std::move(contract));
+        if(!entry.second)
+        {
+          error().source_location = new_symbol.location;
+          error() << "contract '" << new_symbol.display_name()
+                  << "' already set at " << entry.first.location.as_string()
+                  << eom;
+          throw 0;
+        }
+
+        // Remove the contract from the original symbol as we have created a
+        // dedicated contract symbol.
+        new_symbol.type.remove(ID_C_spec_assigns);
+        new_symbol.type.remove(ID_C_spec_ensures);
+        new_symbol.type.remove(ID_C_spec_requires);
+
+        // remove parameter declarations if those were only added for type
+        // checking the contract
+        if(new_symbol.value.is_nil())
+        {
+          for(const auto &parameter : code_type.parameters())
+          {
+            const irep_idt &identifier = parameter.get_identifier();
+
+            symbol_tablet::symbolst::const_iterator p_s_it =
+              symbol_table.symbols.find(identifier);
+            if(p_s_it != symbol_table.symbols.end())
+              symbol_table.erase(p_s_it);
```

Comment:
Here I would rather replace the contract parameters with fresh symbols (maybe just by prefixing the function parameters with a `contract::`  prefix, and also rewrite the contract clauses with these new parameters using a `replacet`. 

Later we would be able to generate a body for the contract without symbol clashes between function and contract symbols and make the contract fully self-contained.

Reply:
I believe we could still choose to create those fresh symbols if and when needed: all the required information still is in the contract symbol.